extensions/ipynb/src/notebookAttachmentCleaner.ts

```diff
@@ -0,0 +1,191 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import { clearTimeout, setTimeout } from 'timers';
+
+/*	Format:
+ *	attachmentCache : { notebookUri : { cellFragment : { filename : { mime : b64 } } } }
+ */
+const attachmentCache: any = {};
+
+class DelayedTrigger implements vscode.Disposable {
+	private timerId: NodeJS.Timeout | undefined;
+
+	/**
+	 * Delay calling the function in callback for a predefined amount of time.
+	 * @param callback : Callback that should be called after some time has passed.
+	 * @param ms : Amount of time after the last trigger that the call to callback
+	 *             should be delayed.
+	 */
+	constructor(
+		private readonly callback: (...args: any[]) => void,
+		private readonly ms: number,
+	) { }
+
+	public trigger(...args: unknown[]): void {
+		if (this.timerId) {
+			clearTimeout(this.timerId);
+		}
+
+		this.timerId = setTimeout(() => {
+			this.callback(...args);
+		}, this.ms);
+	}
+
+	public dispose(): void {
+		if (this.timerId) {
+			clearTimeout(this.timerId);
+		}
+	}
+}
+
+function deleteCacheUri(e: vscode.NotebookDocument) {
+	for (const entry of Object.keys(attachmentCache)) {
+		if (entry === e.uri.toString()) {
+			delete attachmentCache[entry];
+		}
+	}
+}
+
+function renameCacheUri(e: vscode.FileRenameEvent) {
+	const re = /\.ipynb$/;
+	for (const file of e.files) {
+		if (!re.exec(file.oldUri.toString())) {
+			continue;
+		}
+		if (Object.keys(attachmentCache).includes(file.oldUri.toString())) {
+			attachmentCache[file.newUri.toString()] = attachmentCache[file.oldUri.toString()];
+			delete attachmentCache[file.oldUri.toString()];
+		}
+	}
+}
+
+function cleanNotebookAttachments(e: vscode.NotebookDocumentChangeEvent) {
+
+	// return if there are no content changes to the cell source
+	if (!e.cellChanges.length || e.cellChanges[0].document === undefined) {
```

Comment:
What is `e.cellChanges[0]` has no document but the `e.cellChanges[1]` does?

Reply:
Haven't been able to find a situation in which `cellChanges[1]` even exists honestly, but that's an easy fix. I'll shift the `.document` check into the loop of changes and just continue if it isn't there. The `.length` check will remain there.